net: tcp: Remove the 'goto next_state' in tcp_in()

shrek-wang opened the pull request:
Each incoming TCP packet has been completely handled in current state. No need to do further process by 'goto next_state'.

shrek-wang committed:
net: tcp: Remove the 'goto next_state' in tcp_in()

Each incoming TCP packet has been completely handled in current
state. No need to do further process by 'goto next_state'.

Signed-off-by: Shrek Wang <inet_eman@outlook.com>

rlubos reviewed (commented):
I think we still need to do some cleanup with `TCP_CLOSE_WAIT` processing? It used to be transitiory state between `TCP_ESTABLISHED` and `TCP_LAST_ACK` state:
https://github.com/zephyrproject-rtos/zephyr/blob/fb1161a6a90066ff43bb923cdb8639289334bc38/subsys/net/ip/tcp.c#L3318-L3323
But without reiterating the state machine it's just a dead code. I think we could remove the jump to that state in:
https://github.com/zephyrproject-rtos/zephyr/blob/fb1161a6a90066ff43bb923cdb8639289334bc38/subsys/net/ip/tcp.c#L3126
and jump to `TCP_LAST_ACK` unconditionally - in fact, the outcome of processing both `if` blocks is exactly the same, just expressed differently.

shrek-wang commented:
@rlubos, yes, the CLOSE_WAIT part of code is indeed a dead code for now, because it will never go there. Some more parts also need some change in the following steps. Please refer to the topic here for details (any comments are welcome :-) - [https://lists.zephyrproject.org/g/networking/message/312](url)

My rough plan is, to polish the TCP code piece by piece because each commit needs to be specific to 1 feature or issue. Many of the features are relevant. It is really hard for me to find the best sequence to modify which parts first. I'm planning to do the CLOSE_WAIT polish a bit later, after the ACK validation enabled.
For example, this part of code will be also changed, after ACK validation. The 'else' block won't be there.
https://github.com/zephyrproject-rtos/zephyr/blob/024deb49b18519a0a56dfbd53f8c708a496d52f1/subsys/net/ip/tcp.c#L3119-L3127

I'm not sure which way is better - to modify the above 'else' block of code in this commit or later with ACK-validation?

rlubos commented:
I would personally prefer to cleanup any state transitions that relied on the `goto next_state;` statement in this PR to avoid confusion (the only one I could identify is the aforementioned `CLOSE_WAIT`).

shrek-wang commented:
@rlubos, got you. It makes a lot sense. Thanks for your suggestion.
I've modified the code as you suggested. A little difference is that, I removed the var 'acked' so no need the if/else condition now.
